napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -0,0 +1,257 @@
+import numpy as np
+from copy import copy
+from ._shapes_constants import Mode
+from ._shapes_models import Rectangle, Ellipse, Line, Path, Polygon
+from ._shapes_utils import point_to_lines
+
+
+def highlight(layer, event):
+    """Highlight hovered shapes."""
+    layer._set_highlight()
+
+
+def select(layer, event):
+    """Select shapes or vertices either in select or direct select mode.
+
+    Once selected shapes can be moved or resized, and vertices can be moved
+    depending on the mode. Holding shift when resizing a shape will preserve
+    the aspect ratio.
+    """
+    shift = 'Shift' in event.modifiers
+    # on press
+    layer._moving_value = copy(layer._value)
+    if layer._value[1] is None:
+        if shift and layer._value[0] is not None:
+            if layer._value[0] in layer.selected_data:
+                layer.selected_data.remove(layer._value[0])
+                shapes = layer.selected_data
+                layer._selected_box = layer.interaction_box(shapes)
+            else:
+                layer.selected_data.append(layer._value[0])
+                shapes = layer.selected_data
+                layer._selected_box = layer.interaction_box(shapes)
+        elif layer._value[0] is not None:
+            if layer._value[0] not in layer.selected_data:
+                layer.selected_data = [layer._value[0]]
+        else:
+            layer.selected_data = []
+    layer._set_highlight()
+    yield
+
+    # on move
+    while event.type == 'mouse_move':
+        # Drag any selected shapes
+        coord = [layer.coordinates[i] for i in layer.dims.displayed]
+        layer._move(coord)
+        yield
+
+    # on release
+    shift = 'Shift' in event.modifiers
+    if not layer._is_moving and not layer._is_selecting and not shift:
+        if layer._value[0] is not None:
+            layer.selected_data = [layer._value[0]]
+        else:
+            layer.selected_data = []
+    elif layer._is_selecting:
+        layer.selected_data = layer._data_view.shapes_in_box(layer._drag_box)
+        layer._is_selecting = False
+        layer._set_highlight()
+    layer._is_moving = False
+    layer._drag_start = None
+    layer._drag_box = None
+    layer._fixed_vertex = None
+    layer._moving_value = (None, None)
+    layer._set_highlight()
+    layer._update_thumbnail()
+
+
+def add_line_rectangle_ellipse(layer, event):
```

Comment:
I would create a function for each mode and use a helper function.

e.g.
```python
def add_line(layer, event):
    ...
    yield from _add_line_rectangle_ellipse(layer, event, data=np.array([corner, corner + size]), shape_type='line')
```

Reply:
done, though this did mean I had to add some nested if statements inside the shapes model, curious if you prefer it, have ideas on how it can be improved

Reply:
done